### PR TITLE
FP16 training

### PR DESCRIPTION
**Pull request by abhi-iyer**
- Added FP16 option using Torch CUDA AMP
- Added single segment + KW config file for constructing variations on the MT10 experiment
- 1.6x speedup from FP32 to FP16 training

### CODE REVIEW

File: nupic/embodied/multitask/algos/custom_mtsac.py [hunk 1]
```diff
@@ -196,6 +206,226 @@ def train(self, trainer):
...
+                next_observation: :math:`(N, O^*)`
+        Returns:
+            torch.Tensor: loss from the Policy/Actor.
+        """
+        obs = samples_data["observation"]
+
+        with torch.no_grad():
+            alpha = self._get_log_alpha(samples_data).exp()
+
+        with autocast(enabled=self._fp16):
```
Comment (lucasosouza) [inline nupic/embodied/multitask/algos/custom_mtsac.py:302]: Nit: if you can have a hierarchy of contexts, wouldn't be easier to just put autocast prior to line 299 instead of having it here and inside get log alpha?

File: nupic/embodied/multitask/algos/custom_mtsac.py [hunk 2]
```diff
@@ -196,6 +206,226 @@ def train(self, trainer):
...
+        if (log_alpha.shape[0] != one_hots.shape[1]
+                or one_hots.shape[1] != self._num_tasks
+                or log_alpha.shape[0] != self._num_tasks):
+            raise ValueError(
+                "The number of tasks in the environment does "
+                "not match self._num_tasks. Are you sure that you passed "
+                "The correct number of tasks?")
+        
+        with autocast(enabled=self._fp16):
+            return torch.mm(one_hots, log_alpha.unsqueeze(0).t()).squeeze()
```
Comment (lucasosouza) [inline nupic/embodied/multitask/algos/custom_mtsac.py:244]: Do we know for sure we need to autocast this operation as well?

**Review by lucasosouza — APPROVED** [review]
Looks good, great job. Left one nitpick comment and one question, but feel free to merge when you want to.

File: nupic/embodied/multitask/algos/custom_mtsac.py (hunk 2 above)
Comment (abhi-iyer) [reply nupic/embodied/multitask/algos/custom_mtsac.py:244]: Yeah we need it because log alpha is computed _with_ gradients tracked when computing the alpha loss

File: nupic/embodied/multitask/algos/custom_mtsac.py (hunk 1 above)
Comment (abhi-iyer) [reply nupic/embodied/multitask/algos/custom_mtsac.py:302]: Yep I can do this

Comment (abhi-iyer) [reply nupic/embodied/multitask/algos/custom_mtsac.py:302]: Actually, I think it might be cleaner this way - `_get_log_alpha()` already has covered by a autocast context, so it's redundant to do it again. 

Every method should automatically require its components to be covered by an autocast context if FP16 is used, instead of autocasting in some main method where the other functions are called. If I changed it here, I would have to change it for  `_critic_objective` and `_actor_objective`, and it might look messy.

What do you think?

Comment (lucasosouza) [reply nupic/embodied/multitask/algos/custom_mtsac.py:302]: Got it. Yeah, agree with you.
